dimagi/ext/jsonobject.py

```diff
@@ -69,8 +69,8 @@ def _unwrap(self, value):
         return value, safe_strftime(value, '%Y-%m-%dT%H:%M:%S.%fZ')
 
 
-re_trans_datetime = re.compile(r'^\d{4}-0[1-9]|1[0-2]-[12]\d|0[1-9]|3[01]T'
-                               r'[01]\d|2[0-3]:[0-5]\d:[0-5]\d(\.\d{6})?Z$')
+re_trans_datetime = re.compile(r'^(\d{4})-(0[1-9]|1[0-2])-([12]\d|0[1-9]|3[01])T'
+                               r'([01]\d|2[0-3]):([0-5]\d:[0-5]\d)(\.\d{6})?Z$')
```

Comment:
Ok I figured it out:
- `re.match` _always_ starts matching at the beginning (regardless of whether there's a `^`)
- `|` is high precedence, so it was the top-level grouping:
  - `^\d{4}-0[1-9]`
  - `|1[0-2]-[12]\d`
  - `|0[1-9]|3[01]T[01]\d'
  - `|2[0-3]:[0-5]\d:[0-5]\d(\.\d{6})?Z$`
- Since no datetime strings are going to start with any but the first one, when it was matching it was because the string matched (at the front) `^\d{4}-0[1-9]` (so `2015-01-yourmom` would have matched, see below for proof)
- When October came, current datetime strings no longer front-matched that regex


```python
>>> import re
>>> re_trans_datetime = re.compile(r'^\d{4}-0[1-9]|1[0-2]-[12]\d|0[1-9]|3[01]T'
...                                r'[01]\d|2[0-3]:[0-5]\d:[0-5]\d(\.\d{6})?Z$')
>>> re_trans_datetime.match("2015-01-yourmom")
<_sre.SRE_Match at 0x1097cfbe8>
```